src/ray/pubsub/publisher.cc

```diff
@@ -14,22 +14,100 @@
 
 #include "ray/pubsub/publisher.h"
 
+#include "ray/common/ray_config.h"
+
 namespace ray {
 
 namespace pubsub {
 
 namespace pub_internal {
 
-bool SubscriptionIndex::AddEntry(const std::string &key_id,
-                                 const SubscriberID &subscriber_id) {
+bool EntityState::Publish(const rpc::PubMessage &pub_message) {
+  if (subscribers_.empty()) {
+    return false;
+  }
+
+  const int64_t message_size = pub_message.ByteSizeLong();
+
+  while (!pending_messages_.empty()) {
+    // NOTE: if atomic ref counting becomes too expensive, it should be possible
+    // to implement inflight message tracking across subscribers with non-atomic
+    // ref-counting or with a LRU-like data structure tracking the range of buffered
+    // messages for each subscriber.
+    auto front_msg = pending_messages_.front().lock();
```

Comment:
since we store weak_ref in pending_messages_, the total_size_ might be over-counting the actual memory usage. is the understanding correct?

Reply:
Yeah `total_size_` is approximate since we do not count memory used for shared_ptr etc. FWIW, `total_size_` is only used after subtracting the sizes of destroyed messages.